iree/hal/vulkan/api.cc

```diff
@@ -80,58 +81,118 @@ iree_hal_vulkan_syms_release(iree_hal_vulkan_syms_t* syms) {
 //===----------------------------------------------------------------------===//
 
-VulkanDriver::Options ConvertDriverOptions(
-    iree_hal_vulkan_driver_options_t options) {
-  VulkanDriver::Options driver_options;
-  driver_options.api_version = options.api_version;
+namespace {
+
+ExtensibilitySpec GetInstanceExtensibilitySpec(
+    const iree_hal_vulkan_features_t& features) {
+  ExtensibilitySpec spec;
 
-  // TODO: validation layers have bugs when using VK_EXT_debug_report, so if the
-  // user requested that we force them off with a warning. Prefer using
-  // VK_EXT_debug_utils when available.
-  if ((options.extensibility_options & IREE_HAL_VULKAN_ENABLE_DEBUG_REPORT) &&
-      (options.extensibility_options &
-       IREE_HAL_VULKAN_ENABLE_VALIDATION_LAYERS)) {
-    LOG(WARNING) << "VK_EXT_debug_report has issues with modern validation "
-                    "layers; disabling validation";
-    options.extensibility_options =
-        static_cast<iree_hal_vulkan_extensibility_options_t>(
-            options.extensibility_options &
-            ~IREE_HAL_VULKAN_ENABLE_VALIDATION_LAYERS);
+  if (features & IREE_HAL_VULKAN_ENABLE_DEBUG_UTILS) {
+    spec.optional_extensions.push_back(VK_EXT_DEBUG_UTILS_EXTENSION_NAME);
   }
+  if (features & IREE_HAL_VULKAN_ENABLE_PUSH_DESCRIPTORS) {
+    spec.optional_extensions.push_back(
+        VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME);
+  }
+
+  return spec;
+}
+
+ExtensibilitySpec GetDeviceExtensibilitySpec(
+    const iree_hal_vulkan_features_t& features) {
+  ExtensibilitySpec spec;
 
   // REQUIRED: these are required extensions that must be present for IREE to
   // work (such as those relied upon by SPIR-V kernels, etc).
-  driver_options.device_extensibility.required_extensions.push_back(
+  spec.required_extensions.push_back(
       VK_KHR_STORAGE_BUFFER_STORAGE_CLASS_EXTENSION_NAME);
 
-  if (options.extensibility_options &
-      IREE_HAL_VULKAN_ENABLE_VALIDATION_LAYERS) {
-    driver_options.instance_extensibility.optional_layers.push_back(
-        "VK_LAYER_LUNARG_standard_validation");
+  if (features & IREE_HAL_VULKAN_ENABLE_PUSH_DESCRIPTORS) {
+    spec.optional_extensions.push_back(VK_KHR_PUSH_DESCRIPTOR_EXTENSION_NAME);
+  }
+
+  return spec;
+}
+
+}  // namespace
+
+IREE_API_EXPORT iree_status_t IREE_API_CALL iree_hal_vulkan_get_extensions(
+    iree_hal_vulkan_extensibility_set_t extensibility_set,
+    iree_hal_vulkan_features_t features, iree_host_size_t extensions_capacity,
+    const char** out_extensions, iree_host_size_t* out_extensions_count) {
+  if (!out_extensions_count) {
+    return IREE_STATUS_INVALID_ARGUMENT;
   }
+  *out_extensions_count = 0;
 
-  if (options.extensibility_options & IREE_HAL_VULKAN_ENABLE_DEBUG_REPORT) {
-    driver_options.instance_extensibility.optional_extensions.push_back(
-        VK_EXT_DEBUG_REPORT_EXTENSION_NAME);
+  bool is_instance;
+  bool is_required;
+  switch (extensibility_set) {
```

Comment:
if you construct the enum with bits you could then get these bools
INSTANCE_REQUIRED = INSTANCE_BIT | REQUIRED_BIT
bool is_instance = set & INSTANCE_BIT

Reply:
Clever, and I'm not sure if clever is a good thing here :P . Changed to use bits.